fastapi_users/authentication/backend.py

```diff
@@ -0,0 +1,65 @@
+from typing import Any, Callable, Generic
+
+from fastapi import Response
+
+from fastapi_users import models
+from fastapi_users.authentication.strategy import (
+    Strategy,
+    StrategyDestroyNotSupportedError,
+)
+from fastapi_users.authentication.transport import (
+    Transport,
+    TransportLogoutNotSupportedError,
+)
+
+
+class AuthenticationBackend(Generic[models.UC, models.UD]):
+    """
+    Combination of an authentication transport and strategy.
+
+    Together, they provide a full authentication method logic.
+
+    :param name: Name of the backend.
+    :param transport: Authentication transport instance.
+    :param get_strategy: Dependency callable returning
+    an authentication strategy instance.
+    """
+
+    name: str
+    transport: Transport
+
+    def __init__(
+        self,
+        name: str,
+        transport: Transport,
+        get_strategy: Callable[..., Strategy[models.UC, models.UD]],
+    ):
+        self.name = name
+        self.transport = transport
+        self.get_strategy = get_strategy
+
+    async def login(
+        self,
+        strategy: Strategy[models.UC, models.UD],
```

Comment:
Is the `strategy` argument needed, or could it be retrieved from `self.get_strategy()`?

Reply:
Not in the general case: `get_strategy` can be a FastAPI dependency (with sub-dependencies) and, in this case, it should be run inside FastAPI dependency injection context.

This is why `get_strategy` is added as dependency in the login route, and passed manually to this login method:

https://github.com/fastapi-users/fastapi-users/blob/96a086c60f99f28c930460b4c180e6853f12183b/fastapi_users/router/auth.py#L46-L69